clients/src/main/resources/common/message/LeaderAndIsrRequest.json

```diff
@@ -28,59 +28,48 @@
       "about": "The current controller epoch." },
     { "name": "BrokerEpoch", "type": "int64", "versions": "2+", "ignorable": true, "default": "-1",
       "about": "The current broker epoch." },
+    { "name": "PartitionStatesV0", "type": "[]LeaderAndIsrRequestPartition", "versions": "0-1",
+      "about": "The state of each partition, in a v0 or v1 message." },
+    // In v0 or v1 requests, each partition is listed alongside its topic name.
+    // In v2+ requests, partitions are organized by topic, so that each topic name
+    // only needs to be listed once.
     { "name": "TopicStates", "type": "[]LeaderAndIsrRequestTopicState", "versions": "2+",
       "about": "Each topic.", "fields": [
       { "name": "Name", "type": "string", "versions": "2+", "entityType": "topicName",
         "about": "The topic name." },
-      { "name": "PartitionStates", "type": "[]LeaderAndIsrRequestPartitionState", "versions": "0+",
-        "about": "The state of each partition", "fields": [
-        { "name": "PartitionIndex", "type": "int32", "versions": "0+",
-          "about": "The partition index." },
-        { "name": "ControllerEpoch", "type": "int32", "versions": "0+",
-          "about": "The controller epoch." },
-        { "name": "LeaderKey", "type": "int32", "versions": "0+", "entityType": "brokerId",
-          "about": "The broker ID of the leader." },
-        { "name": "LeaderEpoch", "type": "int32", "versions": "0+",
-          "about": "The leader epoch." },
-        { "name": "IsrReplicas", "type": "[]int32", "versions": "0+",
-          "about": "The in-sync replica IDs." },
-        { "name": "ZkVersion", "type": "int32", "versions": "0+",
-          "about": "The ZooKeeper version." },
-        { "name": "Replicas", "type": "[]int32", "versions": "0+",
-          "about": "The replica IDs." },
-        { "name": "IsNew", "type": "bool", "versions": "1+", "default": "false", "ignorable": true, 
-          "about": "Whether the replica should have existed on the broker or not." }
-      ]}
+      { "name": "PartitionStatesV0", "type": "[]LeaderAndIsrRequestPartition", "versions": "2+",
```

Comment:
Looks like we are changing the `name` and the `versions`. Is this a bug in the old schema?

Reply:
Not sure if exactly a bug, but the old schema was definitely more confusing.  `PartitionState` itself is a field in `TopicStates` and since `TopicState` is for versions 2+, this field wouldn't show up in older versions no matter what is the specific version of the field.

Colin unified this field into the new common structure `LeaderAndIsrRequestPartition` which has `TopicName` that only shows up in versions 0-1. So if you have the 2+ `TopicStates` structure, this field won't show up. But when used in  the older (0-1) struct `PartitionStatesV0`, it will be included. This is basically what the new comment above is trying to explain.

Reply:
Thanks for the explanation. That's right.